File: js/particle.js
class Particle {
   constructor(x, y, size, color) {
      this.x = x;
      this.y = y;
      this.size = size;
      this.color = color;
      this.gravity = 0.05;
      this.friction = 0.99;
      this.speedX = Math.random() * 5 - 3;
      this.speedY = Math.random() * 5 - 3;
      this.alpha = Math.random() * 2;
      this.isVisible = false;
   }

   update() {
      this.x += this.speedX;
      this.y += this.speedY;
      this.opacity -= 0.01;
      this.speedX *= this.friction;
      this.speedY *= this.friction;
      this.speedY += this.gravity;
      this.alpha -= 0.02;
      if (this.alpha < 0.02) {
         this.isVisible = true;
         this.alpha = 0;
      }
   }

   draw(ctx) {
      ctx.globalAlpha = this.alpha;
      ctx.fillStyle = this.color;
      ctx.fillRect(this.x, this.y, this.size, this.size);
      ctx.globalAlpha = 1;
   }
}

File: cpp/particle.cpp
#include "./particle.h"
#include "./random.h"

Particle::Particle(float _x, float _y, float _size, short _colorIndex) : x(_x), y(_y), size(_size), colorIndex(_colorIndex) {
   gravity = 0.05f;
   friction = 0.99f;
   alpha = rnd(1.0f, _size / 4);
   vx = rnd(-_size, _size);
   vy = rnd(-_size, _size);
}

void Particle::draw(DrawParticle drawParticle) {
   drawParticle(x, y, size, alpha, colorIndex);
}

bool Particle::update() {
   if (alpha < 0.1f) {
      return true;
   }
   x += vx;
   y += vy;
   vx *= friction;
   vy *= friction;
   vy += gravity;
   alpha *= 0.98f;
   return false;
}
